File: src/main/java/com/github/spafka/concurrent/future/JFuture.java
package com.github.spafka.concurrent.future;

import org.junit.Test;

import java.util.Date;
import java.util.concurrent.*;

public class JFuture {


    @Test
    public void base() {

        System.out.println("start at " + new Date());
        FutureTask<Date> task = new FutureTask<>(() -> {
            System.out.println(Thread.currentThread().getName() + " -> " + new Date());
            Thread.sleep(3000);
            System.out.println(Thread.currentThread().getName() + " -> " + new Date());
            return new Date();
        });
        new Thread(task).start();

        Date date = null;
        try {
            date = task.get(3, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (TimeoutException e) {
            e.printStackTrace();
        }
        System.out.println("end " + date);
    }

    @Test
    public void completeFuture() throws ExecutionException, InterruptedException {

        useCompletableFuture();

        TimeUnit.SECONDS.sleep(1);
    }

    private static void useCompletableFuture() throws InterruptedException, ExecutionException {
        System.out.println("CompletableFuture");

        CompletableFuture.supplyAsync(() -> "A").thenAcceptAsync(name -> {
            System.out.println(name);
        });

    }


}

File: flink/flink-core/src/test/scala/io/github/spafka/rpc/JFuture.scala
package io.github.spafka.rpc

import java.util.function.{Function => JFunction}

import io.github.spafka.util.Logging
import org.junit.Test

class JFuture extends Logging {

  import java.util.concurrent.Executors

  val ex = Executors.newCachedThreadPool()

  @Test def testFuture(): Unit = {
    import java.util.concurrent.{Callable, CompletableFuture, TimeUnit}

    ex.submit(new Callable[Int]() {
      override def call(): Int = 1
    })

    val stringToString: JFunction[_ >: String, _ <: String] = (a: String) ⇒
      a + "c"
    val value: CompletableFuture[String] = CompletableFuture
      .supplyAsync[String](() ⇒ {
      "a"
    })
      .thenApplyAsync[String](((t: String) ⇒ {
      t + "b"
    }): JFunction[_ >: String, _ <: String]) //
      .thenApplyAsync[String](((t: String) ⇒ {
      t + "c"
    }): JFunction[_ >: String, _ <: String]) //
    println(value.get())
    TimeUnit.SECONDS.sleep(1)
  }

}
